### ISSUE TITLE
cert-manager acme challenge fails when gateway exists for host after upgrade to 1.7

### ISSUE THREAD

**Issue by sazzle2611** [istio/istio#27643]
**Bug description**
After upgrading from 1.6.2 to 1.7.2 then cert-manager acme challenges failed to complete. I finally worked out that if I deleted the istio gateway resource for that host then the certificate request completed.

```
apiVersion: cert-manager.io/v1
kind: ClusterIssuer
metadata:
  name: letsencrypt-istio
spec:
  acme:
    server: https://acme-v02.api.letsencrypt.org/directory
    email: hostmaster@example.com
    privateKeySecretRef:
      name: letsencrypt-istio
    solvers:
    - selector: {}
      http01:
        ingress:
          class: istio
```
The order, challenge and ingress are all created and the acme pod starts but it just never completes.

**Affected product area (please put an X in all that apply)**

[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[X] User Experience
[ ] Developer Infrastructure

**Expected behavior**
Certificate requests complete successfully as they did in version 1.6.2

**Version (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version` if you used Helm)**
$ istioctl version --remote
client version: 1.7.2
control plane version: 1.7.2
data plane version: 1.7.2 (57 proxies)

$ kubectl version --short
Client Version: v1.17.3
Server Version: v1.17.3

**How was Istio installed?**
istioctl upgrade -f istio-operator.yaml

**Environment where bug was observed (cloud vendor, OS, etc)**
Kubernetes 1.17.3 on-premise on Centos 7

**Comment by sazzle2611** [comment]
Update

I downgraded to version 1.6.11 and the letsencrypt challenges started succeeding again.

We also use istio multi cluster for certificates from a central vault pod and this wasn't working in version 1.17 either. I didn't mention it before as I thought it was a vault issue but as soon as I downgraded istio then it started working again.

Thanks,
Sarah

**Comment by howardjohn** [comment]
@sazzle2611 can you possibly post the Gateway configuration you have that is conflicting? We do tests this combination in general (cert-manager acme challenge) so it must be specific to the config you are using. Would be good to get a config dump from the gateway as well to avoid back and forth: https://github.com/istio/istio/wiki/Troubleshooting-Istio. Thanks!

**Comment by sazzle2611** [comment]
Hi @howardjohn 

This is the config yaml that we use for one of our k8s clusters (the istio-operator.yaml file that we use to install and update istio)

```
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
spec:
  addonComponents:
    istiocoredns:
      enabled: true
      k8s:
        strategy:
          type: RollingUpdate
          rollingUpdate:
            maxSurge: 1
            maxUnavailable: 1
        hpaSpec:
          minReplicas: 2
          maxReplicas: 3
        affinity:
          podAntiAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
            - labelSelector:
                matchExpressions:
                - key: app
                  operator: In
                  values:
                  - istiocoredns
              topologyKey: kubernetes.io/hostname
            preferredDuringSchedulingIgnoredDuringExecution:
            - weight: 1
              podAffinityTerm:
                labelSelector:
                  matchExpressions:
                  - key: app
                    operator: In
                    values:
                    - istiocoredns
                topologyKey: failure-domain.beta.kubernetes.io/zone
    prometheus:
      enabled: false
    grafana:
      enabled: false
    kiali:
      enabled: false
    tracing:
      enabled: true

  components:
    pilot:
      k8s:
        strategy:
          type: RollingUpdate
          rollingUpdate:
            maxSurge: 1
            maxUnavailable: 1
        hpaSpec:
          minReplicas: 2
          maxReplicas: 3
        affinity:
          podAntiAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
            - labelSelector:
                matchExpressions:
                - key: app
                  operator: In
                  values:
                  - istiod
              topologyKey: kubernetes.io/hostname
            preferredDuringSchedulingIgnoredDuringExecution:
            - weight: 1
              podAffinityTerm:
                labelSelector:
                  matchExpressions:
                  - key: app
                    operator: In
                    values:
                    - istiod
                topologyKey: failure-domain.beta.kubernetes.io/zone
    ingressGateways:
    - name: istio-ingressgateway
      enabled: true
      k8s:
        strategy:
          type: RollingUpdate
          rollingUpdate:
            maxSurge: 1
            maxUnavailable: 1
        hpaSpec:
          minReplicas: 2
          maxReplicas: 3
        affinity:
          podAntiAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
            - labelSelector:
                matchExpressions:
                - key: app
                  operator: In
                  values:
                  - istio-ingressgateway
              topologyKey: kubernetes.io/hostname
            preferredDuringSchedulingIgnoredDuringExecution:
            - weight: 1
              podAffinityTerm:
                labelSelector:
                  matchExpressions:
                  - key: app
                    operator: In
                    values:
                    - istio-ingressgateway
                topologyKey: failure-domain.beta.kubernetes.io/zone
        service:
          ports:
          - name: status-port
            port: 15021
            targetPort: 15021
          - name: http2
            port: 80
            targetPort: 8080
          - name: https
            port: 443
            targetPort: 8443
          - name: tls
            port: 15443
            targetPort: 15443
          - name: tcp-redis
            port: 6379
            targetPort: 6379
          loadBalancerIP: 192.168.28.123
    - name: wx5-ingressgateway
      enabled: true
      label:
        app: wx5-ingressgateway
      k8s:
        strategy:
          type: RollingUpdate
          rollingUpdate:
            maxSurge: 1
            maxUnavailable: 1
        hpaSpec:
          minReplicas: 2
          maxReplicas: 3
        affinity:
          podAntiAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
            - labelSelector:
                matchExpressions:
                - key: app
                  operator: In
                  values:
                  - wx5-ingressgateway
              topologyKey: kubernetes.io/hostname
            preferredDuringSchedulingIgnoredDuringExecution:
            - weight: 1
              podAffinityTerm:
                labelSelector:
                  matchExpressions:
                  - key: app
                    operator: In
                    values:
                    - wx5-ingressgateway
                topologyKey: failure-domain.beta.kubernetes.io/zone
            - weight: 10
              podAffinityTerm:
                labelSelector:
                  matchExpressions:
                  - key: app
                    operator: In
                    values:
                    - wx9-ingressgateway
                    - wx10-ingressgateway
                topologyKey: kubernetes.io/hostname
        service:
          ports:
          - name: status-port
            port: 15021
            targetPort: 15021
          - name: http2
            port: 80
            targetPort: 8080
          - name: https
            port: 443
            targetPort: 8443
          - name: tls
            port: 15443
            targetPort: 15443
          - name: tcp-sftp
            port: 51051
            protocol: TCP
            targetPort: 51051
          loadBalancerIP: 192.168.28.120
          externalTrafficPolicy: Local
    - name: wx9-ingressgateway
      enabled: true
      label:
        app: wx9-ingressgateway
      k8s:
        strategy:
          type: RollingUpdate
          rollingUpdate:
            maxSurge: 1
            maxUnavailable: 1
        hpaSpec:
          minReplicas: 2
          maxReplicas: 3
        affinity:
          podAntiAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
            - labelSelector:
                matchExpressions:
                - key: app
                  operator: In
                  values:
                  - wx9-ingressgateway
              topologyKey: kubernetes.io/hostname
            preferredDuringSchedulingIgnoredDuringExecution:
            - weight: 1
              podAffinityTerm:
                labelSelector:
                  matchExpressions:
                  - key: app
                    operator: In
                    values:
                    - wx9-ingressgateway
                topologyKey: failure-domain.beta.kubernetes.io/zone
            - weight: 10
              podAffinityTerm:
                labelSelector:
                  matchExpressions:
                  - key: app
                    operator: In
                    values:
                    - wx5-ingressgateway
                    - wx10-ingressgateway
                topologyKey: kubernetes.io/hostname
        service:
          ports:
          - name: status-port
            port: 15021
            targetPort: 15021
          - name: http2
            port: 80
            targetPort: 8080
          - name: https
            port: 443
            targetPort: 8443
          - name: tls
            port: 15443
            targetPort: 15443
          - name: tcp-sftp
            port: 51051
            protocol: TCP
            targetPort: 51051
          - name: tcp-sftp-mfi
            port: 51052
            protocol: TCP
            targetPort: 51052
          - name: tcp-gflistener
            port: 10052
            protocol: TCP
            targetPort: 10052
          loadBalancerIP: 192.168.28.128
          externalTrafficPolicy: Local
    - name: wx10-ingressgateway
      enabled: true
      label:
        app: wx10-ingressgateway
      k8s:
        strategy:
          type: RollingUpdate
          rollingUpdate:
            maxSurge: 1
            maxUnavailable: 1
        hpaSpec:
          minReplicas: 2
          maxReplicas: 6
        resources:
          limits:
            cpu: "2"
            memory: 1Gi
          requests:
            cpu: 200m
            memory: 156Mi
        affinity:
          podAntiAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
            - labelSelector:
                matchExpressions:
                - key: app
                  operator: In
                  values:
                  - wx10-ingressgateway
              topologyKey: kubernetes.io/hostname
            preferredDuringSchedulingIgnoredDuringExecution:
            - weight: 1
              podAffinityTerm:
                labelSelector:
                  matchExpressions:
                  - key: app
                    operator: In
                    values:
                    - wx10-ingressgateway
                topologyKey: failure-domain.beta.kubernetes.io/zone
        service:
          ports:
          - name: status-port
            port: 15021
            targetPort: 15021
          - name: http2
            port: 80
            targetPort: 8080
          - name: https
            port: 443
            targetPort: 8443
          - name: tls
            port: 15443
            targetPort: 15443
          - name: tcp-sftp
            port: 51051
            protocol: TCP
            targetPort: 51051
          loadBalancerIP: 192.168.28.122
    egressGateways:
    - name: istio-egressgateway
      enabled: true
      k8s:
        strategy:
          type: RollingUpdate
          rollingUpdate:
            maxSurge: 1
            maxUnavailable: 1
        hpaSpec:
          minReplicas: 2
          maxReplicas: 3
        affinity:
          podAntiAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
            - labelSelector:
                matchExpressions:
                - key: app
                  operator: In
                  values:
                  - istio-egressgateway
              topologyKey: kubernetes.io/hostname
            preferredDuringSchedulingIgnoredDuringExecution:
            - weight: 1
              podAffinityTerm:
                labelSelector:
                  matchExpressions:
                  - key: app
                    operator: In
                    values:
                    - istio-egressgateway
                topologyKey: failure-domain.beta.kubernetes.io/zone
        service:
          ports:
          - name: http2
            port: 80
            targetPort: 8080
          - name: https
            port: 443
            targetPort: 8443
          - name: tls
            port: 15443
            targetPort: 15443

  values:
    global:
      prometheusNamespace: monitoring
      # Provides dns resolution for global services
      podDNSSearchNamespaces:
        - global
      multiCluster:
        enabled: true
      controlPlaneSecurityEnabled: true
    meshConfig:
      accessLogFile: '/dev/stdout'
    sidecarInjectorWebhook:
      neverInjectSelector:
      - matchExpressions:
        - key: app
          operator: In
          values:
          - csi-rbdplugin-provisioner
          - rook-ceph-crashcollector
          - rook-ceph-detect
          - rook-ceph-detect-version
          - rook-ceph-mgr
          - rook-ceph-mon
          - rook-ceph-operator
          - rook-ceph-osd
          - rook-ceph-osd-prepare
          - rook-ceph-tools
          - rook-detect
          - rook-discover
    gateways:
      istio-ingressgateway:
        sds:
          enabled: true
      istio-egressgateway:
        env:
          # Needed to route traffic via egress gateway if desired.
          ISTIO_META_REQUESTED_NETWORK_VIEW: "external"
```

I'm not totally sure what config dump you are asking for. Is it for one our gateways or for the istio-ingressgateway. Also we have downgraded all our clusters back to 1.6 where letsencypt works so are you wanting a dump of a 1.7 gateway?

We have been intending to set up a new k8s cluster so I'm happy to do that this weekend and give you any info that you need if that helps?

Thanks,
Sarah

**Comment by howardjohn** [comment]
> I finally worked out that if I deleted the istio gateway resource for that host then the certificate request completed.
I am looking for the `Gateway` that you needed to delete

**Comment by sazzle2611** [comment]
Ah sorry, this is one of them, from a different k8s cluster but they were all the same

```
apiVersion: networking.istio.io/v1beta1
kind: Gateway
metadata:
  name: tpcoding
  namespace: istio-system
spec:
  selector:
    app: wx11-ingressgateway
    istio: ingressgateway
  servers:
  - hosts:
    - tpcoding.com
    - www.tpcoding.com
    port:
      name: http
      number: 80
      protocol: HTTP
    tls:
      httpsRedirect: true
  - hosts:
    - tpcoding.com
    - www.tpcoding.com
    port:
      name: https
      number: 443
      protocol: HTTPS
    tls:
      mode: SIMPLE
      credentialName: tpcoding-tls
  ```

**Comment by SeanWallace** [comment]
Sorry to bump an old issue but this seems to be the same issue I'm having. On 1.6 cert-manager was able to connect to the ACME challenge pod and on 1.7 and 1.8 it is not.

What I have learned is this seems to come down to `httpsRedirect` being enabled. When I have that enabled and attempt to hit the "well-known" ingress:

```
$ curl -v -HHost:www.mydomain.com "192.168.1.250/.well-known/acme-challenge/21RZvQxYINfJcl8HBOnU92IpALjh05XrRs357F71LhQ"
*   Trying 192.168.1.250...
* TCP_NODELAY set
* Connected to 192.168.1.250 (192.168.1.250) port 80 (#0)
> GET /.well-known/acme-challenge/21RZvQxYINfJcl8HBOnU92IpALjh05XrRs357F71LhQ HTTP/1.1
> Host:www.mydomain.com
> User-Agent: curl/7.58.0
> Accept: */*
>
< HTTP/1.1 301 Moved Permanently
< location: https://www.mydomain.com/.well-known/acme-challenge/21RZvQxYINfJcl8HBOnU92IpALjh05XrRs357F71LhQ
< date: Sun, 03 Jan 2021 00:09:25 GMT
< server: istio-envoy
< content-length: 0
```

Once I update the gateway spec I get this:
```
$ curl -v -HHost:www.mydomain.com "192.168.1.250/.well-known/acme-challenge/21RZvQxYINfJcl8HBOnU92IpALjh05XrRs357F71LhQ"
*   Trying 192.168.1.250...
* TCP_NODELAY set
* Connected to 192.168.1.250 (192.168.1.250) port 80 (#0)
> GET /.well-known/acme-challenge/21RZvQxYINfJcl8HBOnU92IpALjh05XrRs357F71LhQ HTTP/1.1
> Host:www.mydomain.com
> User-Agent: curl/7.58.0
> Accept: */*
>
< HTTP/1.1 200 OK
< cache-control: no-cache, no-store, must-revalidate
< date: Sun, 03 Jan 2021 00:11:31 GMT
< content-length: 87
< content-type: text/plain; charset=utf-8
< x-envoy-upstream-service-time: 1
< server: istio-envoy
```

As a workaround I have been toggling that flag on all my gateways when I need to renew. Given I have a lot of domains this has been a pretty serious regression for me.

**Comment by howardjohn** [comment]
@SeanWallace what Gateway configuration do you have? Can you post the exact thing? Working on a fix for this and want to make sure i fully understand the problem

**Comment by howardjohn** [comment]
My understanding is in 1.6 it was actually broken such that when cert-manager created the Ingress for the challenge, httpsRedirect actually stopped working for *all* requests. In 1.7, this was "fixed", and now it requires TLS on all requests to `www.mydomain.com`. I think the distinction here is we would want the redirect only on things other than `.well-known/acme-challenge/...` but I am not sure Envoy supports that currently

**Comment by howardjohn** [comment]
I think to do it we would need to stop using RequireTLS in envoy, and instead use route_action.https_redirect and a match on `:scheme`. This is not only a big change, but its broken in envoy currently: https://github.com/envoyproxy/envoy/issues/14587. So would likely be a while until fixed.

**Comment by SeanWallace** [comment]
Your theory sounds correct to me just based on my observation.

Here is my fully Gateway config:

```
apiVersion: networking.istio.io/v1beta1
kind: Gateway
metadata:
  name: website-mydomain-com
  namespace: mine
spec:
  selector:
    istio: istio-ingressgateway-mine
  servers:
  - hosts:
    - mydomain.com
    - www.mydomain.com
    port:
      name: http
      number: 80
      protocol: HTTP
    tls:
      httpsRedirect: true
  - hosts:
    - mydomain.com
    - www.mydomain.com
    port:
      name: https
      number: 443
      protocol: HTTPS
    tls:
      credentialName: website-mydomain-com-cert
      minProtocolVersion: TLSV1_2
      mode: SIMPLE
      privateKey: sds
      serverCertificate: sds
```

And in case it's useful the ingress that was generated by cert-manager:

```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  annotations:
    kubernetes.io/ingress.class: istio
    nginx.ingress.kubernetes.io/whitelist-source-range: 0.0.0.0/0,::/0
  creationTimestamp: "2021-01-04T02:08:56Z"
  generateName: cm-acme-http-solver-
  generation: 1
  labels:
    acme.cert-manager.io/http-domain: "2332887661"
    acme.cert-manager.io/http-token: "964497066"
    acme.cert-manager.io/http01-solver: "true"
spec:
  rules:
  - host: mydomain.com
    http:
      paths:
      - backend:
          service:
            name: cm-acme-http-solver-pslzv
            port:
              number: 8089
        path: /.well-known/acme-challenge/pUB07WsWON9TLYcACp-jpLwLnxBKNQZmvIywbfjTC8U
        pathType: ImplementationSpecific
status:
  loadBalancer:
    ingress:
    - ip: 192.168.1.250
```

Thank you for your attention on this, it is much appreciated!

**Comment by SeanWallace** [comment]
I know this issue is old and likely not high priority but can it be kept open? Seems as though this issue was reported again not too long ago.

**Comment by howardjohn** [comment]
Definitely. This is actually not low priority in my opinion, but its really hard to fix unfortunately

**Comment by simonherbert** [comment]
For anyone looking for a solution, here might be one.
This has also be mentioned in https://discuss.istio.io/t/how-to-redirect-to-https-except-for-well-known-acme-challenge/9011/4.
```
apiVersion: cert-manager.io/v1alpha2
kind: Certificate
metadata:
  name: my-cert
  namespace: istio-system
  annotations:
    cert-manager.io/issue-temporary-certificate: "true"
```
Enabling `cert-manager.io/issue-temporary-certificate` generates a self-signed certificate until a valid one is available.
Let's Encrypt states that they follow HTTPS redirects *without* checking the certificate.

> Our implementation of the HTTP-01 challenge follows redirects, up to 10 redirects deep. It only accepts redirects to “http:” or “https:”, and only to ports 80 or 443. It does not accept redirects to IP addresses. When redirected to an HTTPS URL, it does not validate certificates (since this challenge is intended to bootstrap valid certificates, it may encounter self-signed or expired certificates along the way).
Source: https://cert-manager.io/docs/usage/certificate/#temporary-certificates-whilst-issuing

**Comment by camelpunch** [comment]
@simonherbert I was under the impression that this is a different problem: that httpsRedirect used to have a special case to allow cert-manager traffic to its special verification pods, but no longer does?

**Comment by simonherbert** [comment]
@camelpunch They are related. Issuing a temporary self-signed certificate makes the https port available before a certificate could be issued. This might not work with all certificate authorities, but at least Let's Encrypt states that they accept that.
The special case for `.well-known` is the real solution.

**Comment by mway-niels** [comment]
> For anyone looking for a solution, here might be one.
> This has also be mentioned in https://discuss.istio.io/t/how-to-redirect-to-https-except-for-well-known-acme-challenge/9011/4.
> 
> apiVersion: cert-manager.io/v1alpha2
> kind: Certificate
> metadata:
>   name: my-cert
>   namespace: istio-system
>   annotations:
>     cert-manager.io/issue-temporary-certificate: "true"
> 
> Enabling cert-manager.io/issue-temporary-certificate generates a self-signed certificate until a valid one is available.
> Let's Encrypt states that they follow HTTPS redirects without checking the certificate.

I tried this solution, unfortunately Let's Encrypt or Istio (not exactly sure which one) didn't replace the temporary certificate with a valid one. This of course means the browser still displays a warning that the served certificate is self signed and will block the connection by default.

Is there another option/configuration I have to set to get this solution to work?

**Comment by howardjohn** [comment]
With https://github.com/istio/api/pull/2088 you can do this if/when its merged. Not as good of a UX though:

```PR
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: http-redirect
  namespace: default
spec:
  gateways:
  - https-redirect
  hosts:
  - 'mydomain.com'
  http:
  - redirect:
      scheme: https
      redirectCode: 302
---
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: https-redirect
spec:
  selector:
    istio: ingressgateway
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
    hosts:
    - "mydomain.com"
---
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: https
spec:
  selector:
    istio: ingressgateway
  servers:
  - port:
      number: 443
      name: https
      protocol: HTTPS
    hosts:
    - "mydomain.com"
    tls:
      credentialName: sds-credential
      mode: SIMPLE
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: echo
spec:
  hosts:
  - "mydomain.com"
  gateways:
  - https
  http:
  - route:
    - destination:
        host: echo
        port:
          number: 80

```

**Comment by dahovey** [comment]
@howardjohn Is there any updates on this issue? I would rather not split VS and gateway into four resources as you show above if update is on its way.

Thanks.

**Comment by coldguy101** [comment]
Just ran into this problem myself. My work-around was to stop using
```
tls: {
    httpsRedirect: true
},
```
in my Istio Gateway spec, and instead use a VirtualService matcher to do the redirection.
```
    {
      // This rule exists because of the Istio issue on the httpsRedirect flag above
      match: [
        {
          port: 80,
          uri: {
            prefix: "/"
          },
        },
      ],
      redirect: {
        scheme: "https"
      },
    },
...
```
Doing this allowed cert-manager to insert its own HTTP routes, and when those get cleaned up automatically, revert back to the standard redirect.

**Comment by RobCannon** [comment]
@coldguy101 

Thank you for your post.  It didn't quite work for me, but it got me to this solution which is working for me.  Here is the complete declaration.

```
apiVersion: networking.istio.io/v1beta1
kind: Gateway
metadata:
  name: kiali
  namespace: monitoring
spec:
  selector:
    istio: ingressgateway
  servers:
    - port:
        number: 80
        name: http
        protocol: HTTP
      hosts:
        - kiali.sandbox.acceleratelearning.com
    - port:
        number: 443
        name: https
        protocol: HTTPS
      tls:
        mode: SIMPLE
        credentialName: kiali-certificate
      hosts:
        - kiali.sandbox.acceleratelearning.com
---
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: kiali
  namespace: monitoring
spec:
  hosts:
    - "*"
  gateways:
    - kiali
  http:
    # The redirect for https has to happen in the virtual service,
    # otherwise it will prevent the cert-manager http01 challenge from working
    - match:
        - scheme:
            exact: http
      redirect:
        scheme: https
    - route:
        - destination:
            host: kiali
            port:
              number: 20001
```

**Comment by yekinabud** [comment]
@RobCannon you saved my life

Just had to upgrade my Istio control plane from 1.11 to 1.13.4, otherwise the `redirect.scheme` field was broken

**Comment by yekibud** [comment]
@RobCannon you saved my life

Just had to upgrade my Istio control plane from 1.11 to 1.13.4, otherwise the redirect.scheme field was broken

**Comment by sneko** [comment]
**--- BELOW A POTENTIAL SIMPLE HACKY SOLUTION :) ---**

Every 3 months I'm like struggling hard with Istio and CertManager and I'm so sad after years it's still a real pain even for really simple use cases (using `httpsRedirect: true` for example). Hope a proper solution will come out one day!

My solution:
* As stated above, we have to do the HTTPS redirection on the VirtualService level since the gateway cannot specify rules of matching
* As I specified 3 months ago https://github.com/istio/istio/issues/27643#issuecomment-1095306328 the idea is to do the HTTPS redirection on all URIs except for `/.well-known/acme-challenge/` (due to limitation of the RE2 regex It's a bit hacky but working!)

_gateway.yaml_
```yaml
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: main-gateway
  namespace: mynamespace
spec:
  selector:
    istio: ingressgateway
  servers:
    - port:
        number: 80
        name: http-ingress
        protocol: HTTP2
      hosts:
        - "mynamespace/www.mydomain.com"
      tls:
        httpsRedirect: false # The HTTPS redirections will be handled in the VirtualService layer to not conflict with certificates renewing...
    - port:
        number: 443
        name: https-ingress
        protocol: HTTPS
      tls:
        mode: SIMPLE
        privateKey: sds
        serverCertificate: sds
        credentialName: ingress-cert
      hosts:
        - "mynamespace/www.mydomain.com"
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: workaround-to-renew-certs
  namespace: mynamespace
spec:
  hosts:
    - "*"
  gateways:
    - main-gateway
  http:
    # The redirect for https has to happen in the virtual service,
    # otherwise it will prevent the cert-manager http01 challenge from working
    - match:
        - scheme:
            exact: http
          uri:
            # This is not a perfect regex but RE2 does not allow negations so we have to hack a it
            # Note: this pattern will redirect to HTTPS all URLs NOT starting with `/.well-` (which should be the best case since your business URLs are unlikely to start with that)
            # this to let pass requests to renew certs like `/.well-known/acme-challenge/...`
            regex: "^/(([^\\.].*)|(\\.[^w].*)|(\\.w[^e].*)|(\\.we[^l].*)|(\\.wel[^l].*)|(\\.well[^\\-].*))"
      redirect:
        scheme: https
        redirectCode: 302 # 302 to not mess with risky permanent redirections since a workaround
```

🏎️💨

**Comment by howardjohn** [comment]
instead of negation can you just match everything? Then when cert manager creates a rule it will be more specific and take precedence. I haven't tested but I think it could work

**Comment by RobCannon** [comment]
That has been my experience.  The cert-manager rule is inserted dynamically and takes precedence.  But, I tried to build a cluster this morning and the cert-manager challenge is not working, so maybe something has changed somewhere since yesterday.

**Comment by sneko** [comment]
@howardjohn I looked at precedence from the Kubernetes perspective... there is no explicit priority way when dealing with distinct CR, also I only found one guy mentioning it was relying on `creationTimestamp` order...

Same for Istio.

From what I understand you mean there is a logic somewhere that detects rules and say "if stricter than others, top priority". Can you clarify on what's it's based? Is it a logic inside `match:` where `exact (high priority) > prefix (medium) > regex (low)`?

Note: I know multiple rules in a VirtualService are from top to bottom... but since CertManager manages another file... it does not apply here

**Comment by mick1627** [comment]
I had this issue plus the fact that I have several ingress gateway. This two point was resolved using the HTTP-01 Gateway API solver https://cert-manager.io/docs/configuration/acme/http01/#configuring-the-http-01-gateway-api-solver.  The switch from 
istio Gateway,VirtualService to API Gateway and HTTPRoute is easy on our case as we don't have complex VS.

**Comment by sudomann** [comment]
To supplement @sneko 's answer above, we can do a better match for http01 requests.

FROM:
>```yaml
>   http:
>     # The redirect for https has to happen in the virtual service,
>     # otherwise it will prevent the cert-manager http01 challenge from working
>     - match:
>         - scheme:
>             exact: http
>           uri:
>             # This is not a perfect regex but RE2 does not allow negations so we have to hack a it
>             # Note: this pattern will redirect to HTTPS all URIs NOT starting with `/.well-` (which should be the best case since your business URIs are unlikely to start with that)
>             # this to let pass requests to renew certs like `/.well-known/acme-challenge/...`
>             regex: "^/(([^\\.].*)|(\\.[^w].*)|(\\.w[^e].*)|(\\.we[^l].*)|(\\.wel[^l].*)|(\\.well[^\\-].*))"
> ```

TO:
```yaml
  http:
    - match:
        - withoutHeaders:
            :path:
              prefix: /.well-known/acme-challenge/
```

**Comment by romanlum** [comment]
Is there any progress here?
Is the above workaround working correctly?

**Comment by radekg** [comment]
> This worked for us, we changed tls httpsRedirect to false (or commented it out) in the gateway and then added this as the first rule under http in the virtual service
> 
> ```
>   - match:
>       - scheme:
>           exact: http
>     redirect:
>       scheme: https
> ```
> 
> We have so far changed about 50 configs to this in the last 6 months and not had a problem yet

This is the correct solution. I've written a bit more in details about why and how it works: https://gruchalski.com/posts/2023-07-09-istio-cert-manager-lets-encrypt-and-https-redirect/.

**Comment by melkishengue** [comment]
does anyone know how the solution with the virtual service could be implemented in a setup including OAuth2-proxy? The issue right now is that the authorization check comes before the virtual service so there is no redirect to https when the authorization page is displayed. Only once the user has passed the authorization, the redirect happens correctly

**Comment by fritterhoff** [comment]
> ```yaml
>   http:
>     - match:
>         - withoutHeaders:
>             :path:
>               prefix: /.well-known/acme-challenge/
> ```

Sadly that works no longer with the latest release.

istio is complaining:

```
VirtualService/istio-ingress/workaround-to-renew-certs dry-run failed: admission webhook "validation.istio.io" denied the request: configuration is invalid: header name :path is not a valid header name
```

**Comment by sudomann** [comment]
@fritterhoff I don't use Istio currently so I can't check, but does replacing `:path` with `uri` work?

**Comment by fritterhoff** [comment]
Let me check...

**Comment by fritterhoff** [comment]
So replacing path with uri does not work directly.

```
❯ curl -v http://pki.hm.edu/.well-known/acme-challenge/test
* Host pki.hm.edu:80 was resolved.
* IPv6: 2001:4ca0:6800:3::6
* IPv4: 141.39.241.102
*   Trying [2001:4ca0:6800:3::6]:80...
* Connected to pki.hm.edu (2001:4ca0:6800:3::6) port 80
> GET /.well-known/acme-challenge/test HTTP/1.1
> Host: pki.hm.edu
> User-Agent: curl/8.7.1
> Accept: */*
>
* Request completely sent off
< HTTP/1.1 302 Found
< location: https://pki.hm.edu/.well-known/acme-challenge/test
< content-security-policy: frame-ancestors 'none';
< x-frame-options: deny
< x-xss-protection: 1; mode=block
< x-content-type-options: nosniff
< referrer-policy: no-referrer
```

**Comment by sudomann** [comment]
> > instead of negation can you just match everything? Then when cert manager creates a rule it will be more specific and take precedence. I haven't tested but I think it could work
> 
> This worked for us, we changed tls httpsRedirect to false (or commented it out) in the gateway and then added this as the first rule under http in the virtual service
> 
> ```
>   - match:
>       - scheme:
>           exact: http
>     redirect:
>       scheme: https
> ```
> 
> We have so far changed about 50 configs to this in the last 6 months and not had a problem yet

I hope @sazzle2611 's answer still works. It's likely I'll run into this issue again if I deploy issue for a future project

**Comment by fritterhoff** [comment]
I'll check that later.

**Comment by fritterhoff** [comment]
Seems like that could work if I add that fragment into every service...
